src/app/modules/app/pages/classrooms/components/classroom-form/classroom-form.component.ts

```diff
@@ -0,0 +1,194 @@
+
+export type ScheduleForm = FormGroup<{
+  weekday: FormControl<string | null>;
+  startingTime: FormControl<string | null>;
+  endingTime: FormControl<string | null>;
+}>;
+
+@Component({
+  selector: 'app-classroom-form',
+  templateUrl: './classroom-form.component.html',
+  styleUrls: ['./classroom-form.component.sass'],
+  providers: [
+    {
+      provide: MAT_FORM_FIELD_DEFAULT_OPTIONS,
+      useValue: { appearance: 'outline', floatLabel: 'always' },
+    },
+  ],
+})
+export class ClassroomFormComponent implements OnInit {
+  title!: string;
+  classroomForm!: FormGroup;
+  programOptions: SelectOption[] = [];
+  courseOptions: SelectOption[] = [];
+  weekdayOptions = weekdayOptions;
+
+  constructor(
+    public readonly classroomService: ClassroomService,
+    public readonly programService: ProgramService,
+    public readonly courseService: CourseService,
+    private readonly formBuilder: FormBuilder,
+    private readonly dialogRef: MatDialogRef<ClassroomFormComponent>,
+    @Inject(MAT_DIALOG_DATA) public data: { item: Classroom }
+  ) {}
+
+  private initializeForm(): void {
+    this.classroomForm = this.formBuilder.group({
+      course: [
+        this.data ? this.data.item.courseObject?.id?.toString() : '',
+        [Validators.required],
+      ],
+      program: [
+        this.data ? this.data.item.programObject?.id.toString() : '',
+        [Validators.required],
+      ],
+      startingDate: [
+        this.data ? this.data.item.starts : '',
+        [Validators.required],
+      ],
+      schedules: new FormArray<ScheduleForm>([]),
+    });
+  }
+
+  get schedules(): FormArray {
+    return this.classroomForm.get('schedules') as FormArray<ScheduleForm>;
+  }
+
+  addSchedule() {
+    const schedule: ScheduleForm = this.formBuilder.group({
+      weekday: ['', [Validators.required]],
+      startingTime: ['', [Validators.required]],
+      endingTime: ['', [Validators.required]],
+    });
+
+    this.schedules.push(schedule);
+  }
+
+  get scheduleControls(): FormGroup[] {
+    return this.schedules.controls as FormGroup[];
+  }
+
+  ngOnInit(): void {
+    // IMPROVEMENT: this data should be managed in a stream serivce of courses and programs
+    this.fetchCourseOptions();
+    this.fetchProgramOptions();
+
+    this.title = this.data ? 'Edit Classroom' : 'Create Classroom';
+    this.initializeForm();
+  }
+
+  fetchCourseOptions() {
+    this.courseService
+      .getAll()
+      .pipe(
+        map((courses): SelectOption[] =>
+          courses.map(course => ({
+            label: course.name,
+            value: course.id!.toString(),
+          }))
+        )
+      )
+      .subscribe(options => {
+        this.courseOptions = options;
+      });
+  }
+
+  fetchProgramOptions() {
+    this.programService
+      .getAll()
+      .pipe(
+        map((programs): SelectOption[] =>
+          programs.map(program => ({
+            label: program.name,
+            value: program.id.toString(),
```

Comment:
we can implement destructuring  in program variable

Reply:
Updated in last commit